Spawn antifreeze

themrdemonized opened the pull request:
V3
* restored spawn_antifreeze functionality with edits, default is disabled
* Prefetch models in a separate thread before spawning in the main thread
*

themrdemonized committed:
* restored spawn_antifreeze functionality with edits, default is disabled
* When ProcessGameEvents iteration encounter spawn event, it is moved into spawn_events queue
* Separate ProcessSpawnEvents delegate runs in Device.seqParallel to handle spawns in separate thread to some extent
* spawn_antifreeze_batch cvar to handle spawning entities in batches, describes how many spawn events to fire each frame, default 2

themrdemonized committed:
less verbose logging

themrdemonized commented:
Test release https://github.com/themrdemonized/xray-monolith/releases/tag/2025.6.17

Toggle with `spawn_antifreeze 1` in console or in Modded Exes options

themrdemonized committed:
small optimizations

themrdemonized committed:
increase batch to 3

fridvp commented:
had an issue with it
playing with mags redux and quite a lot of other mods. it seems to not spawn all the items it should (as in the items that spawn under same conditions with spawn_antifreeze 0) (talking about adding items to a trader). Also just gives a lot of microfreezes while spawning those items

ProfLander commented:
I played around with this in an attempt to get some thread profiling data, points of note:

- The new loading logic didn't seem to kick in when opening Sid's trade interface on a fresh Loner game, possibly due to items populating before / independently of the antifreeze system, but doing the same with Petrenko on a Duty start reliably triggers it.

- Visual Studio 2022's Concurrency Profiler extension worked once, producing a browseable thread graph, and then disappeared from the extensions menu forever; enable / disable / uninstall / reinstall do nothing. With any luck that's specific to my machine, since it seems to be path-of-least-resistance for threaded tracing.

- As an alternative, I tried AMD uProf, but it doesn't display the advertised per-thread flamegraph when attached to Anomaly; no related options in the UI, built-in presets make no difference, and the manual doesn't specify the preconditions for making it gather / process the necessary data to show it.

We may well end up needing IXRay-style Optick or some other integrated solution for consistent threaded profiling...

themrdemonized commented:
https://github.com/bombomby/optick bomboclat
https://github.com/ixray-team/ixray-1.6-stcop/commit/16e5dacc81515700094004e4665f05150212953d + example integration

themrdemonized committed:
spawn antifreeze v2:
* Instead of device.seqParallel, use thread_spawn
* Prefetch models in separate thread before spawning game object, then spawn them normally
* Removed spawn_antifreeze batch variable, all events come and go same way as before

themrdemonized committed:
removed batch from modded exes oprions, cleanup

themrdemonized committed:
removed parent_id check in ProcessSpawnEvents, the queue is sorted before processing

themrdemonized commented:
V2:
https://github.com/themrdemonized/xray-monolith/releases/tag/2025.6.19

* Instead of device.seqParallel, use thread_spawn
* Prefetch models in separate thread before spawning game object, then spawn them normally
* Removed spawn_antifreeze batch variable, all events come and go same way as before

The idea is to reduce stutters by prefetching models and textures in the separate thread before spawning. I removed batch processing, so scripts and other parts of engine should work correctly as before. Trader items might get freezes back, but i think its bearable

themrdemonized committed:
optimization, instead of persistent thread working all the time, launch new ones consecutively

ProfLander committed:
Integrate optick profiler

ProfLander commented:
Bomboclaat actioned: #262 

Should help with figuring this out!

themrdemonized committed:
don't prefetch models twice

themrdemonized committed:
Revert "don't prefetch models twice"

This reverts commit 2ede4300ebbbcb9272b0c51714c130a9a4cd54be.

ProfLander committed:
Add github action to initialize submodules

ProfLander committed:
Add `git submodule update --init --recursive` step to build instructions

ProfLander committed:
Predefine `OPTICK_ENABLE_GPU_VULKAN=0`

ProfLander committed:
Set `optick` to build for `Verified` / `VerifiedDX11`

themrdemonized committed:
cleanup

themrdemonized committed:
optionally read `spawn_antifreeze_ignore` section, which might contain section names or model names to not use in prefetching

ProfLander committed:
Formalize profiler setup, disable by default

themrdemonized committed:
don't prefetch models twice

themrdemonized committed:
Merge pull request #262 from ProfLander/optick

Integrate optick profiler

themrdemonized committed:
cleanup

themrdemonized committed:
ProfiledDX11 configuration

themrdemonized committed:
fixed ProfiledDX11 config, xrPhysics dont have Profiling yet, Link error when trying to use Profiled config

themrdemonized committed:
Fixed build output for optick lib

themrdemonized committed:
Merge branch 'optick' into spawn-antifreeze

themrdemonized committed:
Add prefetching of actor inventory in actor_on_first_update in script

themrdemonized committed:
replaced trans_outfit.script that did 65534 loop with better

themrdemonized committed:
restore permanent thread with handle

themrdemonized commented:
![изображение](https://github.com/user-attachments/assets/f95cd201-cc38-4290-b701-c496f10afc53)
So there is a noticeable freeze that happens 2 frames before game object spawn kicks in and subsequent prefetching. The freeze happens when trying to spawn an NPC in particular While prefetching does save some time, its not the main cause for a freeze. Need investigating

themrdemonized committed:
Add profiler points

themrdemonized committed:
refactor to use process_queue vector type

themrdemonized committed:
add multiple visuals per section
remove prefetched models registry for now

themrdemonized committed:
Read actual alife visual to prefetch

themrdemonized commented:
V3: https://github.com/themrdemonized/xray-monolith/releases/tag/2025.6.21

This one looks pretty good, works with actual visuals of alife objects. Need to test for consistency and stability.

As the sidenote of this feature, any scripts that spawn alife objects and expect game_object on next frame won't work correctly. It usually was the case for vanilla code as well but its guaranteed with this feature.

themrdemonized committed:
elaborativity refactor
remove model filepath extension before prefetching

themrdemonized commented:
![изображение](https://github.com/user-attachments/assets/1c3869a0-fe0e-4e84-8c4a-db01d5e1fc87)
`army_sim_squad_10` squad
900 ms loading stutter is offloaded, actual spawning is 93 ms

themrdemonized committed:
Revert "remove duplicate stack traces on Lua-related CTD"

This reverts commit fded13d382405cf012c08996d4910c9f8c32092f.

themrdemonized committed:
profile save events

themrdemonized committed:
replace sort with stable_sort for spawn queue

themrdemonized committed:
discord in separate thread

themrdemonized committed:
fix dynamic manager loot table containing possible destroyed objects

themrdemonized committed:
trans_outfit.transparent_gg() 65534 loop fix

themrdemonized committed:
Options: Optimizations page out of debug group

themrdemonized committed:
solution file gamedata fix

themrdemonized committed:
trans_outfit.transparent_gg() 65534 loop fix

themrdemonized committed:
changed psLUA_GCSTEP 400 -> 80

themrdemonized committed:
Migration script to update cvars and other stuff on existing installations

themrdemonized committed:
gcStep to 50

themrdemonized committed:
replace some printf to printf_me

themrdemonized committed:
after tests, too low lua_gc lead to occasional big stutters, increase to 160

themrdemonized committed:
moving files out of vs2022

themrdemonized committed:
solution file p1

themrdemonized committed:
sdk files p1

themrdemonized committed:
sdk files

themrdemonized committed:
src files

themrdemonized committed:
src files p2

themrdemonized committed:
luabind

LVutner commented:
peak modding

themrdemonized committed:
Spawn Antifreeze turn on by default

themrdemonized committed:
Merge branch 'vs2022-fix-pch' into spawn-antifreeze

themrdemonized committed:
changelog
